### ISSUE TITLE
Automatically clear stale widget states

### ISSUE THREAD

**Issue by philippbc** [jupyter-widgets/ipywidgets#2598]
I have a simple Jupyter notebook with an ipywidget.Label. If I run the cell containing the Label twice, the widget states of the first and second run are _both_ kept in the notebook state universe. Is this intended? If I understand it correctly, the widget from the first run has disappeared and so its state information is stale. What is the purpose of keeping it?

I've found a few related issues: #2012 and #2565

Wouldn't this be a more intuitive approach:
- only ever keep state information of widgets that are currently displayed (from a previous and current kernels);
- the option to save states should be persistent (i.e., I should not need to remember every time I update a widget to save the states) --> #2565.

**Comment by vidartf** [comment]
There are multiple cases were you would want to keep a widget around even if it is not displayed:
- You are creating a set of widgets that you want to alternate between showing depending on user action, e.g. "show theses widgets if user pressed button A, but show these other widgets if user pressed button B".
- having pure widget models as part of a hierarchy data that can be displayed, e.g. like how pythreejs does
- etc.

Having automatic saving of the state should now be possible with the lab extension I think?

**Comment by philippbc** [comment]
Right, those make sense. So what would be the easiest way to clean up the state space when re-running cells without clearing the states of live widgets?

**Comment by annelisagrout** [comment]
You can "close" widgets explicitly: `w.close()`. I think we don't have a good way to automatically clean up state, short of restarting the kernel and refreshing the browser page, which clears all state.

**Comment by jasongrout** [comment]
You can explicitly close a widget: "widget.close()". I think we don't have a good way to automatically clear widget state (because it's not clear when we should clear state) short of refreshing your browser and restarting the kernel, which clears all state.

**Comment by maartenbreddels** [comment]
That's why I added `ipywidgets.Widget.close_all()`  :), although there are some issues with it now: #2564 
I hope it does what you want.

**Comment by pbcornelius** [comment]
@maartenbreddels If I understand `close_all()` correctly, it also closes live/displayed widgets?

@jasongrout Is it possible to register with the widget's cell so that one can invoke `.close()` when the cell is cleared? Like a callback?

**Comment by maartenbreddels** [comment]
> @maartenbreddels If I understand `close_all()` correctly, it also closes live/displayed widgets?

Yes, we don't know how many are visible, unless we set `_view_count` from the start to `0`, but we don't. 

@pbcornelius #3108 might solve your problem, in a slightly different way (I think)

**Comment by stevejpurves** [comment]
i think this is also related to issue [#3007](https://github.com/jupyter-widgets/ipywidgets/issues/3007) which seemed to also have different behaviour in lab than in the classic notebook
